Blog on Super-TDMA (#54)
org-arl/unet-blog: _posts/2020-07-06-whats-so-super-about-super-tdma.md

---
layout: post
comments: true

> title: What's so "super" about Super-TDMA ?

Reviewer: While the title is _catchy_, might we benefit from a mention of UnetStack being able to do it somehow? To think about...

Author: The title I used went very well with the flow of this blog in my opinion but I agree that mention of UnetStack may help. I have these three titles in mind. Please do suggest your preference on these below-mentioned titles or let me know if you have think of a different title.

1. What's so "super" about Super-TDMA ?
2. Simulating protocols exploiting long propagation delays in UnetSim
3. Super-TDMA : Key concepts, practical challenges and it's simulation in UnetSim

Reviewer: 1 still is best :-)

Author: Ok :) I will stick with this title then.

date: 06/07/2020
author: Prasad Anjangi
categories: howto
feature-img: "assets/img/stdma/stdma-unet.jpg"
thumbnail: "assets/img/stdms/stdma-unet.jpg"
tags: [howto, tdma, medium access control, long propagation delays, throughput, performance boost]
---

> Needless to say that underwater acoustic (UWA) networks might play a key role in many areas including marine, offshore and subsea industries in the future. There have been tremendous and impressive technological advances in underwater acoustic communications and networking field. The one that caught my attention and brought about a clear difference in the novel approaches that are needed to deal with challenges in UWA networks was the "exploitation" of large propagation delays that exist in UWA networks. Do you know that it is possible to achieve as much as 50% higher network throughput with specific network geometries and protocols in UWA networks when compared to radio-frequency (RF) based terrestrial wireless networks? It is an interesting and surprising fact that has led to further exploration of such techniques in practical settings. In this blog, without delving too much into the nitty-gritty details, let's try and understand; where that advantage comes from, through an example technique (*Super-TDMA*). We will try and understand the key idea behind this technique  and let's see why it is "super".

Author: Fixed.

![Overview](../assets/img/stdma/stdma-schedule-2.png)
This can be easily visualized [2][3]. Take a look at the figure below where the above-mentioned schedule is repeated for 12-time slots or 3 periods. The line segments in black represent transmissions whereas the line segments in blue represent receptions. The line segments in dotted red are the interfering packets. It is interesting to see that most of these interfering packets are aligned in a time slot and such time slots are utilized for transmissions. This is an essential principle using which algorithms can be developed for much more complex network geometries. This idea opens up tremendous possibilities in designing practical techniques for medium access control in networks with large propagation delays.

### Can we simulate such a schedule in UnetSim ?

> Yes, of course, let's do that next. It is easy to implement a 3 node equilateral network geometry with the above-mentioned high throughput schedule in UnetSim. The sample code as `e3-network.groovy `is provided here. Let us go through the code a bit to try and understand how it is implemented.

Reviewer: Might be good to break the code into parts and explain each (either before or after showing the complete code).

> Yes, of course, let's do that next. It is easy to implement a 3 node equilateral network geometry with the above-mentioned high throughput schedule in UnetSim. The sample code as `e3-network.groovy `is provided here. Let us go through the code a bit to try and understand how it is implemented.

Author: Yes, I tested it and the code works. I have now explained relevant code snippets.

RX:                     ${trace.rxCount}
Offered load:           ${trace.offeredLoad.round(3)}
Throughput:             ${trace.throughput.round(3)}"""
```

First we set the distance between the nodes to be 650 m. With sound speed in water to be approximately 1540 m/s, the slot length turns out to be 422 ms which is set as can be seen in the code. The simulation time is set to 15 minutes and the high-throughput schedule is given as an input. The rest is pretty simple. The three network nodes are deployed at appropriate coordinates to form an equilateral triangle. The frame duration of the DATA channel packet is set to be smaller than the time slot length, i.e., we need to make sure that the frame length and data rate is such that the packet duration is smaller than 422 ms. Once we set this up, the packets are transmitted from each node in a `TickerBehavior` as per the schedule. Now, let's run this and look at the result:

```term
Equilateral triangle network
----------------------------
Internode distance:     650 m
Slot length:            422 ms
Simulation time:        900 s
TX:                     3198
RX:                     3011
Offered load:           1.375

> Throughput:             1.295

Reviewer: Might want some brief explanation on how to interpret `throughput`, and what throughput > 1 means (this is what we call _super_).

Author: Agreed. I have added much more details on this part now.

1 simulation completed in 2.615 seconds
```

> The network capacity as we learned earlier is 1.5 (see [1] for more details). Although, in this network, our offered load is less than that since the packet duration is slightly lesser than the time slot duration and hence they are not fully utilized. And therefore, we see a reduced throughput but still 29.5% larger than what can be achieved traditionally. The fact that the practical frame durations used in underwater networks are comparable to the propagation delays led to this gain in throughput. This is remarkable and we should aspire to harness such knowledge of propagation delays in networks when designing performant protocols for higher throughput.

Reviewer: There's a bunch of assumptions about synchronization and clocks. It would seem naïve to assume perfect sync and not address how a practical system might work, at least at a conceptual level.

Author: I've now added common practical issues that need to be taken care of and some ways in which they might be addressed.

---
layout: post
comments: true
title: What's so "super" about Super-TDMA ? 
date: 06/07/2020
author: Prasad Anjangi
categories: howto
feature-img: "assets/img/stdma/stdma-unet.jpg"
thumbnail: "assets/img/stdms/stdma-unet.jpg"
tags: [howto, tdma, medium access control, long propagation delays, throughput, performance boost]
---

> Needless to say that underwater acoustic (UWA) networks might play a key role in many areas including marine, offshore and subsea industries in the future. There have been tremendous and impressive technological advances in underwater acoustic communications and networking field. The one that caught my attention and brought about a clear difference in the novel approaches that are needed to deal with challenges in UWA networks was the "exploitation" of large propagation delays that exist in UWA networks. Do you know that it is possible to achieve as much as 50% higher network throughput with specific network geometries and protocols in UWA networks when compared to radio-frequency (RF) based terrestrial wireless networks? It is an interesting and surprising fact that has led to further exploration of such techniques in practical settings. In this blog, without delving too much into the nitty-gritty details, let's try and understand, where that advantage comes from, through an example technique (*Super-TDMA*). We will try and understand the key idea behind this technique  and let's see why it is "super".

Reviewer 2: ```suggestion
Underwater acoustic (UWA) networks will play a key role in many areas including marine, offshore and subsea industries in the future. There have been tremendous and impressive technological advances in underwater acoustic communications and networking field. The one that caught my attention and brought about a clear difference in the novel approaches that are needed to deal with challenges in UWA networks was the "exploitation" of large propagation delays that exist in UWA networks. 

Do you know that it is possible to achieve as much as 50% higher network throughput with specific network geometries and protocols in UWA networks when compared to radio-frequency (RF) based terrestrial wireless networks? 

It is an interesting and surprising fact that has led to further exploration of such techniques in practical settings. In this blog, let's try and understand, where that advantage comes from, through an example technique (*Super-TDMA*). We will try and understand the key idea behind this technique and let's see why it is "super".
```

"50% higher network throughput" <- would this be normalized throughput? Without the context this sounds like we're 50% faster than a terrestrial network.

Author: Yes you are right. Modified the statement now to reflect that correctly.

### Long propagation delays

![Overview](../assets/img/stdma/stdma-lpd.png)

> Let us first understand what is meant by a long propagation delay. Specifically, how "long" it has to be? If the propagation delay in a medium is larger (or in the same order) in comparison to the message duration, i.e., the time it takes when a signal/message is transmitted is much larger in comparison to the signal/message duration, then the propagation delay is considered long. This is quite easy to understand if we look at an example. Consider a transmitter and a receiver node placed at say 750 m apart. When placed in water, an acoustic signal carrying information will take approximately 500 ms to reach the receiver, whereas, in air, an RF signal for the same setup will take approximately 2.5 microseconds. If the message duration is in the order of milliseconds, it is easy to see that the long propagation delay exists in the UWA setup where the propagation delay of 500 ms is much larger.

Reviewer 2: ```suggestion
Let us first understand what is meant by a long propagation delay. Specifically, how "long" does it have to be? If the propagation delay in a medium is larger (or in the same order) in comparison to the message duration, i.e., the time it takes when a signal/message is transmitted is much larger in comparison to the signal/message duration, then the propagation delay is considered long. This is quite easy to understand if we look at an example. Consider a transmitter and a receiver node placed at say 750 m apart. When placed in water, an acoustic signal carrying information will take approximately 500 ms to reach the receiver, whereas, in air, an RF signal for the same setup will take approximately 2.5 microseconds. If the message duration is in the order of milliseconds, it is easy to see that the long propagation delay exists in the UWA setup where the propagation delay of 500 ms is much larger.
```

### Traditional TDMA

> It is easiest to show how the long propagation delay can be harnessed to achieve higher throughput using a simple contention-free medium access control technique called Time Division Multiple Access (TDMA). In traditional TDMA protocols, every node in the network is assigned a fixed time slot in which it can transmit information. Appropriate amounts of guard time in each time slot must be left to make sure the intended receiver node receives the message successfully. As you can imagine, this protocol may not be great in networks that suffer from long propagation delays, right ?

Reviewer 2: ```suggestion
It is easiest to show how the long propagation delay can be harnessed to achieve higher throughput using a simple contention-free medium access control technique called [Time Division Multiple Access (TDMA)](https://en.wikipedia.org/wiki/Time-division_multiple_access). In traditional TDMA protocols, every node in the network is assigned a fixed time slot in which it can transmit information. Appropriate amounts of guard time in each time slot must be left to make sure the intended receiver node receives the message successfully. As you can imagine, this protocol may not be great in networks that suffer from long propagation delays, right ?
```

---
layout: post
comments: true
title: What's so "super" about Super-TDMA ? 
date: 06/07/2020
author: Prasad Anjangi
categories: howto
feature-img: "assets/img/stdma/stdma-unet.jpg"
thumbnail: "assets/img/stdms/stdma-unet.jpg"
tags: [howto, tdma, medium access control, long propagation delays, throughput, performance boost]
---

> Underwater acoustic (UWA) networks will play a key role in many areas including marine, offshore and subsea industries in the future. There have been tremendous and impressive technological advances in underwater acoustic communications and networking field. The one that caught my attention and brought about a clear difference in the novel approaches that are needed to deal with challenges in UWA networks was the "exploitation" of large propagation delays that exist in UWA networks. 

Reviewer 3: I would say UWA is already playing a key role now :) not in the future!

Author: Agreed :) Fixed now.

### Long propagation delays

![Overview](../assets/img/stdma/stdma-lpd.png)

Let us first understand what is meant by a long propagation delay. Specifically, how "long" does it have to be? If the propagation delay in a medium is larger (or in the same order) in comparison to the message duration, i.e., the time it takes when a signal/message is transmitted is much larger in comparison to the signal/message duration, then the propagation delay is considered long. This is quite easy to understand if we look at an example. Consider a transmitter and a receiver node placed at say 750 m apart. When placed in water, an acoustic signal carrying information will take approximately 500 ms to reach the receiver, whereas, in air, an RF signal for the same setup will take approximately 2.5 microseconds. If the message duration is in the order of milliseconds, it is easy to see that the long propagation delay exists in the UWA setup where the propagation delay of 500 ms is much larger.

### Traditional TDMA

> It is easiest to show how the long propagation delay can be harnessed to achieve higher throughput using a simple contention-free medium access control technique called [Time Division Multiple Access (TDMA)](https://en.wikipedia.org/wiki/Time-division_multiple_access). In traditional TDMA protocols, every node in the network is assigned a fixed time slot in which it can transmit information. Appropriate amounts of guard time in each time slot must be left to make sure the intended receiver node receives the message successfully. As you can imagine, this protocol may not be great in networks that suffer from long propagation delays, right ?

Reviewer 3: ```suggestion
It is easiest to show how the long propagation delay can be harnessed to achieve higher throughput using a simple contention-free medium access control technique called [Time Division Multiple Access (TDMA)](https://en.wikipedia.org/wiki/Time-division_multiple_access). In traditional TDMA protocols, every node in the network is assigned a fixed time slot in which it can transmit information. Appropriate amounts of guard time in each time slot must be left to make sure the intended receiver node receives the message successfully. As you can imagine, this protocol may not be great in networks that suffer from long propagation delays, right?
```

![Overview](../assets/img/stdma/stdma-ttdma.png)
Consider a large 10 km network, i.e., the farthest distance between a transmitter and receiver node is 10 km. In this case, for a message to reach the farthest receiving node, it may take approximately ~ 7 seconds (see above figure). If the message duration is 1 second long, the channel utilization efficiency is 1/8 which is 12.5% and not very impressive.

### Super TDMA

The key idea involved in Super-TDMA is visualized in the figure below. Since the message duration is much shorter than the propagation delay, a natural question arises. 

*Can a message be scheduled to be transmitted in one slot and be received in a different slot ?* 

If the transmissions and receptions are carefully scheduled to occur in different slots, it turns out that much higher throughput is achievable.

![Overview](../assets/img/stdma/stdma-stdma.png)

> Note that in traditional TDMA protocol a significant amount of time in each slot is left vacant, however, in protocols employing the exploitation of large propagation delays, such large guard times are not needed.

Reviewer 3: ```suggestion
Note that in traditional TDMA protocol, a significant amount of time in each slot is left vacant. However, in protocols employing the exploitation of large propagation delays, such large guard times are not needed.
```

----------------------------
Internode distance:     ${range} m
Slot length:            ${(1000*slot).round()} ms
Simulation time:        ${time} s"""

///////////////////////////////////////////////////////////////////////////////
// simulate schedule

simulate time, {

  def n = []
  n << node('1', address: 1, location: [0, 0, 0])
  n << node('2', address: 2, location: [range, 0, 0])
  n << node('3', address: 3, location: [0.5*range, 0.866*range, 0])

  n.eachWithIndex { n1, i ->
    n1.startup = {
      def phy = agentForService PHYSICAL
      phy[Physical.DATA].frameLength = phy[Physical.CONTROL].frameLength
      add new TickerBehavior(1000*slot, {
        def slen = schedule[i].size()
        def s = schedule[i][(tickCount-1)%slen]
        if (s) phy << new TxFrameReq(to: s, type: Physical.DATA)
      })
    }
  }

}

// display statistics
println """TX:                     ${trace.txCount}
RX:                     ${trace.rxCount}
Offered load:           ${trace.offeredLoad.round(3)}
Throughput:             ${trace.throughput.round(3)}"""
```
Let us go through some of the code snippets from the above script to try and understand different components of the implementation.

> First we set the distance between the nodes to be 650 m. With sound speed in water to be approximately 1540 m/s, the propagation delay on any link turns out to be 423 ms. Therefore, the slot duration is set to 423 ms. The simulation time is set to 15 minutes and the high-throughput schedule structured as a matrix is given as an input to all three nodes in the network. 

Reviewer 3: If we do not indicate shallow/well mixed/isovelocity, this may come across as odd, since many readers may be familiar with deep waters where taking sound speed profile is the norm.

Author: Agreed. But, such information is generally abstracted out for network simulations as is the case here. To not digress from the topic at hand, my opinion is to skip this explanation. Let me know otherwise.

Reviewer 3: If it is normal to abstract this in simulations (as long as it is safe to assume our audience knows it), let's go with this.
